Client/core/FPSLimiter.cpp

```diff
@@ -0,0 +1,399 @@
+    }
+
+#pragma endregion Interface
+
+#pragma region Internal
+
+    std::uint32_t FPSLimiter::GetDisplayRefreshRate()
+    {
+        D3DDISPLAYMODE DisplayMode;
+        IDirect3D9*    pD3D9 = CProxyDirect3D9::StaticGetDirect3D();
+        if (pD3D9 && pD3D9->GetAdapterDisplayMode(D3DADAPTER_DEFAULT, &DisplayMode) == D3D_OK)
+        {
+            return ValidateFPS(DisplayMode.RefreshRate);
+        }
+        // Fallback to 60 Hz if unable to get refresh rate
+        return 60;
+    }
+
+    // Priority: Server > Client > User > None
+    void FPSLimiter::CalculateCurrentFPSLimit()
+    {
+        OutputDebugLine("FPSLimiter: Calculating current FPS limit...");
+        std::uint32_t oldLimit = m_data.activeFPSTarget;
+
+        if (m_data.hasPendingLimit)
+        {
+            m_data.activeFPSTarget = m_data.pendingFPSTarget;
+            m_data.hasPendingLimit = false;
+        }
+
+        // Find the minimum FPS among all enforced limits (excluding unlimited and below minimum)
+        std::uint32_t minFPS = std::numeric_limits<std::uint32_t>::max();
+
+        if (m_data.serverEnforcedFPS > FPS_LIMIT_MIN && m_data.serverEnforcedFPS < minFPS)
+            minFPS = m_data.serverEnforcedFPS;
+        if (m_data.clientEnforcedFPS > FPS_LIMIT_MIN && m_data.clientEnforcedFPS < minFPS)
+            minFPS = m_data.clientEnforcedFPS;
+        if (m_data.userDefinedFPS > FPS_LIMIT_MIN && m_data.userDefinedFPS < minFPS)
+            minFPS = m_data.userDefinedFPS;
+        if (m_data.displayRefreshRate > FPS_LIMIT_MIN && m_data.displayRefreshRate < minFPS)
+            minFPS = m_data.displayRefreshRate;
+
+        if (minFPS == std::numeric_limits<std::uint32_t>::max())
+            minFPS = FPS_LIMIT_UNLIMITED;
+
+        m_data.activeFPSTarget = minFPS;
+
+        // If limit changed, reset frame timing
+        if (oldLimit != m_data.activeFPSTarget)
+        {
+            QueryPerformanceCounter(&m_lastFrameTime);
+            m_data.appliedThisFrame = false;
+
+            // Call FPS limit change handler
+            OnFPSLimitChange();
+        }
+
+        std::stringstream ss;
+        ss << "FPSLimiter: FPS limit changed to " << (m_data.activeFPSTarget == 0 ? "unlimited" : std::to_string(m_data.activeFPSTarget));
+        ss << " (Enforced by: " << EnumToString(GetEnforcer()) << ")";
+
+        auto* pConsole = CCore::GetSingleton().GetConsole();
+        if (pConsole)
+            CCore::GetSingleton().GetConsole()->Print(ss.str().c_str());
+
+        OutputDebugLine(ss.str().c_str());
+    }
+
+    void FPSLimiter::SetFrameRateThrottle()
+    {
+        TIMING_GRAPH("LimiterStart");
+        auto defer = [&]() { TIMING_GRAPH("LimiterEnd"); };
+
+        if (m_data.activeFPSTarget == FPS_LIMIT_UNLIMITED)
+        {
+            return defer();
+        }
+        else if (m_data.appliedThisFrame)
+        {
+            OutputDebugLine("Warning: FPSLimiter::SetFrameRateThrottle called when already applied this frame");
```

Comment:
If this kind of information is needed, shouldn’t it be displayed only in the debug build? I mean all outputDebugLine in this PR